src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.Unix.cs

```diff
@@ -255,7 +255,7 @@ internal DateTimeOffset GetCreationTime(ReadOnlySpan<char> path, bool continueOn
             return UnixTimeToDateTimeOffset(_fileCache.CTime, _fileCache.CTimeNsec);
         }
 
-        internal void SetCreationTime(string path, DateTimeOffset time)
+        private void SetCreationTime_StandardUnixImpl(string path, DateTimeOffset time)
```

Comment:
Suggestion on function naming and splitting across these files:

-- FileStatus.Unix.cs
```cs
SetAccessOrWriteTime(string path, DateTimeOffset time, bool isAccessTime)
{
  // cache stuff
  SetAccessOrWriteTimeCore(path, time, isAccessTime);
  // cache stuff
}

SetAccessOrWriteTimeImpl(string path, DateTimeOffset time, bool isAccessTime, bool checkCreationTime)
{
..
   if (checkCreationTime && ...)
   {
   	SetCreationTimeCore(path, time);
   }
...
}
```
-- FileStatus.SetTimes.OSX.cs
```cs
SetCreationTime(string path, DateTimeOffset time)
{
  // cache stuff
    SetCreationTimeCore(...)
    -> fall back to calling SetAccessOrWriteTimeCore()
  // cache stuff
}

SetAccessOrWriteTimeCore(string path, DateTimeOffset time, bool isAccessTime)
	=> SetAccessOrWriteTimeImpl(path, time, isAccessTime, checkCreationTime: true);
	
SetCreationTimeCore(string path, DateTimeOffset time)
{
  ...
}
```

-- FileStatus.SetTimes.OtherUnix.cs
```cs
SetCreationTime(string path, DateTimeOffset time)
 => SetWriteTime(path, time);
 
 SetAccessOrWriteTimeCore(string path, DateTimeOffset time, bool isAccessTime)
 => SetAccessOrWriteTimeImpl(path, time, isAccessTime, checkCreationTime: false);

SetCreationTimeCore(string path, DateTimeOffset time)
{ // Not supported. }
```

Reply:
Does this seem reasonable. I've attached it to avoid a rerun of automated tests if it needs changes.
Note: I've intentionally left `SetCreationTimeStandardUnix` in `FileStatus.Unix.cs` because I can't think of a way to do it without code duplication, or without reducing performance (by unnecessarily invalidating at the start of the function call and it would also be unclear because it would lack the explanatory comment).
[Archive.zip](https://github.com/dotnet/runtime/files/7135801/Archive.zip)

Reply:
This has consistent naming `Xxx` calls `XxxCore` (calls `XxxImpl`).
The top-level function `Xxx` is responsible for the cache stuff.

`SetAccessOrWriteTimeImpl` is like `SetCreationTimeStandardUnix` but it includes the logic for updating the creation time, by calling out to `SetCreationTimeCore`.

`OtherUnix.cs` doesn't have any code, so there is no duplication.

There is no reduced performance I am aware of.

Reply:
> `SetAccessOrWriteTimeImpl` is like `SetCreationTimeStandardUnix` but it includes the logic for updating the creation time, by calling out to `SetCreationTimeCore`.

I'm confused by this comment.

The reduced performance is in relation to if 
```cs
 InvalidateCaches();
EnsureCachesInitialized(path);
```
was put at the top of `SetCreationTime`.
Since `EnsureCachesInitialized` is not trivial (IO calls should never be considered trivial), if it can be avoided it should be (which it can).

Reply:
Yes, it should be called before falling back to `SetAccessOrWriteTimeCore`.

Reply:
It doesn't fall back to `SetAccessOrWriteTimeCore`, it falls back to `SetCreationTimeStandardUnix` (called directly on other platforms) which does cache stuff and then calls `SetAccessOrWriteTimeImpl`.

Reply:
It can't call `SetAccessOrWriteTimeCore` because `SetAccessOrWriteTimeCore` can call `SetCreationTimeCore`. This is why there is a seperate `SetCreationTimeStandardUnix` method.

Reply:
In fact, the line `SetCreationTimeCore(path, creationTime);` (I know it is missing the `Core` part on the one I uploaded, a line I missed earlier) (in SetAccessOrWriteTimeCore of OSX) should probably be
```cs
Interop.Error result = SetCreationTimeImpl(path, time);
if (result != Interop.Error.ENOTSUP && result != Interop.Error.SUCCESS)
{
    Interop.CheckIo(result, path, InitiallyDirectory);
}
```
instead.

Reply:
> It can't call SetAccessOrWriteTimeCore because SetAccessOrWriteTimeCore can call SetCreationTimeCore. This is why there is a seperate SetCreationTimeStandardUnix method.

Right, it should call `SetAccessOrWriteTimeImpl(path, time, isAccessTime, checkCreationTime: false);`

Reply:
What is `checkCreationTime`?
It does call `SetAccessOrWriteTimeImpl` through `SetCreationTimeStandardUnix`. It is called through this to be clear about why it is calling `SetAccessOrWriteTimeImpl`. I didn't put that comment there (it was in the old `SetCreationTime`), so I don't think it should be skipped if someone is trying to follow how the code works. If I don't have the `SetCreationTimeStandardUnix` it will lose the clarity why it is calling `SetAccessOrWriteTimeImpl` and will require multiple copies of the same method which also invalidates and refreshes the cache first before the call to `SetAccessOrWriteTimeImpl`.
I'm not really sure what you're suggesting here, but it seems like it would duplicate code and/or cause confusion about why `SetCreationTime` calls `SetAccessOrWriteTimeImpl`.

Reply:
Also, I think `SetAccessOrWriteTimeImpl` with `checkCreationTime = true` will allow to more efficiently implement `(isAccessTime ? GetLastWriteTime(path) : time)` because we know have the last write time, while `GetLastWriteTime(path)` needs to fetch it again since the cache was invalidated.

Reply:
I don't think the cache is invalidated there since `EnsureCachesInitialized` always validates it. And there is no call to `InvalidateCaches` between `EnsureCachesInitialized` and `GetLastWriteTime` (in SetAccessOrWriteTimeCore of OSX implementation) that I can see. So this change would not make it more efficient.

Reply:
This is actually the reason why there is `isAccessTime ? GetLastWriteTime(path) : time` since the time won't be updated if it was setting the write time yet through `GetLastWriteTime`.

Reply:
> What is checkCreationTime?

It merges `SetCreationTime_StandardUnixImpl` and `SetAccessOrWriteTime` into a single shared method, and only the OSX implementation uses `true` to cause the creation time update. 

Based on our discussion, I think the above suggestion can be simplified to:

-- FileStatus.Unix.cs
```cs
SetAccessOrWriteTimeCore(string path, DateTimeOffset time, bool isAccessTime, bool checkCreationTime)
{
  // cache stuff: refresh
  ...
   if (checkCreationTime && <use cached timestamps>)
   {
   	SetCreationTimeCore(path, time);
   }
  // cache stuff: invalidate
}

```
-- FileStatus.SetTimes.OSX.cs
```cs
SetCreationTime(string path, DateTimeOffset time)
{
    SetCreationTimeCore(...)
    ENOTSUP: -> fall back to calling SetAccessOrWriteTimeCore(path, time, isAccessTime: false, checkCreationTime: false);
}

SetAccessOrWriteTime(string path, DateTimeOffset time, bool isAccessTime)
	=> SetAccessOrWriteTimeCore(path, time, isAccessTime, checkCreationTime: true);
	
SetCreationTimeCore(string path, DateTimeOffset time)
{
  ...
  // cache stuff: invalidate
}
```

-- FileStatus.SetTimes.OtherUnix.cs
```cs
SetCreationTime(string path, DateTimeOffset time)
 => SetWriteTime(path, time);
 
SetAccessOrWriteTime(string path, DateTimeOffset time, bool isAccessTime)
	=> SetAccessOrWriteTimeCore(path, time, isAccessTime, checkCreationTime: false);

SetCreationTimeCore(string path, DateTimeOffset time)
{ // Not supported. }
```

> I don't think the cache is invalidated there since EnsureCachesInitialized always validates it. And there is no call to InvalidateCaches between EnsureCachesInitialized and GetLastWriteTime (in SetAccessOrWriteTimeCore of OSX implementation) that I can see. So this change would not make it more efficient.

I see, that's because `SetAccessOrWriteTimeImpl` no longer invalidates the cache.
It would be more clear a cached value is used this wasn't split across methods.

Reply:
I wasn't the one to split the caching and reading of cached values :)
I'll look at these changes now.

Reply:
Hey, I'm not sure how this allows removing `SetCreationTimeStandardUnix` since the purpose of this method is to explain why it sets the write time instead of the creation time and now why the caches are being refreshed as well. And the reason that it has to exist (and not be a part of another method) is because it is used in 2 places. I've attached the newest version of the files.
[Archive.zip](https://github.com/dotnet/runtime/files/7136517/Archive.zip)
It's night time where I live and I'm going to bed now, feel free to leave any further suggestions.

Reply:
> Edit: Also, having SetCreationTimeStandardUnix be seperate won't reduce performance since it should be inlined by the JIT anyway.

We should mostly care about the syscalls. Other stuff like inlineability won't be noticeable because the syscalls are far more costly than what we're doing in .NET code.

> I'm not sure how this allows removing SetCreationTimeStandardUnix since the purpose of this method is to explain why it sets the write time instead of the creation time and now why the caches are being refreshed as well. And the reason that it has to exist (and not be a part of another method) is because it is used in 2 places. I've attached the newest version of the files.

Something like this:

```diff
diff --git a/FileStatus.SetTimes.OSX.cs b/FileStatus.SetTimes.OSX.cs
index 657fda0..f67b5e2 100644
--- a/FileStatus.SetTimes.OSX.cs
+++ b/FileStatus.SetTimes.OSX.cs
@@ -7,7 +7,9 @@ namespace System.IO
 {
     internal partial struct FileStatus
     {
-        private void SetCreationTimeCore(string path, DateTimeOffset time, bool allowFallbackToLastWriteTime = true)
+        static bool s_creationTimeNotSupported;
+
+        private void SetCreationTime(string path, DateTimeOffset time)
         {
             // allowFallbackToLastWriteTime is is included so that when called from
             // SetAccessOrWriteTimeCore, it doesn't try to restore a creation time
@@ -19,11 +21,11 @@ namespace System.IO
             // setattrlist()", so we fall back to the method used on other unix
             // platforms, otherwise we throw an error if we get one, or invalidate
             // the cache if successful because otherwise it has invalid information.
-            Interop.Error result = SetCreationTimeImpl(path, time);
+            Interop.Error result = SetCreationTimeCore(path, time);
 
             if (result == Interop.Error.ENOTSUP)
             {
-                if (allowFallbackToLastWriteTime) SetCreationTimeStandardUnix(path, time);
+                SetLastWriteTime(path, time);
             }
             else if (result != Interop.Error.SUCCESS)
             {
@@ -31,8 +33,16 @@ namespace System.IO
             }
         }
 
-        private Interop.Error SetCreationTimeImpl(string path, DateTimeOffset time)
+        private void SetAccessOrWriteTime(string path, DateTimeOffset time, bool isAccessTime)
+            => SetAccessOrWriteTimeCore(path, time, isAccessTime, updateCreationTime: !s_creationTimeNotSupported);
+
+        private Interop.Error SetCreationTimeCore(string path, DateTimeOffset time)
         {
+            if (s_creationTimeNotSupported)
+            {
+                return Interop.Error.ENOTSUP;
+            }
+
             Interop.Sys.TimeSpec timeSpec = default;
 
             long seconds = time.ToUnixTimeSeconds();
@@ -45,9 +55,16 @@ namespace System.IO
             attrList.bitmapCount = Interop.libc.AttrList.ATTR_BIT_MAP_COUNT;
             attrList.commonAttr = Interop.libc.AttrList.ATTR_CMN_CRTIME;
 
-            return (Interop.Error)Interop.libc.setattrlist(path, &attrList, &timeSpec, sizeof(Interop.Sys.TimeSpec), default(CULong));
-        }
+            Interop.Error error = (Interop.Error)Interop.libc.setattrlist(path, &attrList, &timeSpec, sizeof(Interop.Sys.TimeSpec), default(CULong));
 
-        private void SetAccessOrWriteTimeCore(string path, DateTimeOffset time, bool isAccessTime) => SetAccessOrWriteTimeImpl(path, time, isAccessTime, checkCreationTime: true);
+            s_creationTimeNotSupported = error == Interop.Error.ENOTSUP;
+
+            if (error == Interop.Error.SUCCESS)
+            {
+                InvalidateCaches();
+            }
+
+            return error;
+        }
     }
 }
diff --git a/FileStatus.SetTimes.OtherUnix.cs b/FileStatus.SetTimes.OtherUnix.cs
index db4e921..ad1b473 100644
--- a/FileStatus.SetTimes.OtherUnix.cs
+++ b/FileStatus.SetTimes.OtherUnix.cs
@@ -7,7 +7,10 @@ namespace System.IO
 {
     internal partial struct FileStatus
     {
-        private void SetCreationTimeCore(string path, DateTimeOffset time, bool allowFallbackToLastWriteTime) => SetCreationTimeStandardUnix(path, time);
-        private void SetAccessOrWriteTimeCore(string path, DateTimeOffset time, bool isAccessTime) => SetAccessOrWriteTimeImpl(path, time, isAccessTime, checkCreationTime: false);
+        private void SetCreationTime(string path, DateTimeOffset time)
+            => SetWriteTime(path, time);
+
+        private void SetAccessOrWriteTime(string path, DateTimeOffset time, bool isAccessTime)
+            => SetAccessOrWriteTimeCore(path, time, isAccessTime, updateCreationTime: false);
     }
 }
diff --git a/FileStatus.Unix.cs b/FileStatus.Unix.cs
index 4bf4bf6..a2b838f 100644
--- a/FileStatus.Unix.cs
+++ b/FileStatus.Unix.cs
@@ -255,27 +255,6 @@ namespace System.IO
             return UnixTimeToDateTimeOffset(_fileCache.CTime, _fileCache.CTimeNsec);
         }
 
-        private void SetCreationTimeStandardUnix(string path, DateTimeOffset time)
-        {
-            // This method has the non-OSX implementation for setting a creation date, it is included in this
-            // file rather than the OtherUnix file as it is used as a fallback on OSX.
-
-            // force a refresh so that we have an up-to-date times for values not being overwritten
-            InvalidateCaches();
-            EnsureCachesInitialized(path);
-
-            // Unix provides APIs to update the last access time (atime) and last modification time (mtime).
-            // There is no API to update the CreationTime.
-            // Some platforms (e.g. Linux) don't store a creation time. On those platforms, the creation time
-            // is synthesized as the oldest of last status change time (ctime) and last modification time (mtime).
-            // We update the LastWriteTime (mtime).
-            // This triggers a metadata change for FileSystemWatcher NotifyFilters.CreationTime.
-            // Updating the mtime, causes the ctime to be set to 'now'. So, on platforms that don't store a
-            // CreationTime, GetCreationTime will return the value that was previously set (when that value
-            // wasn't in the future).
-            SetAccessOrWriteTimeImpl(path, time, isAccessTime: false, checkCreationTime: false);
-        }
-
         internal DateTimeOffset GetLastAccessTime(ReadOnlySpan<char> path, bool continueOnError = false)
         {
             EnsureCachesInitialized(path, continueOnError);
@@ -301,41 +280,12 @@ namespace System.IO
             return DateTimeOffset.FromUnixTimeSeconds(seconds).AddTicks(nanoseconds / NanosecondsPerTick).ToLocalTime();
         }
 
-        private void SetAccessOrWriteTime(string path, DateTimeOffset time, bool isAccessTime)
+        private unsafe void SetAccessOrWriteTimeCore(string path, DateTimeOffset time, bool isAccessTime, bool checkCreationTime)
         {
             // force a refresh so that we have an up-to-date times for values not being overwritten
             InvalidateCaches();
             EnsureCachesInitialized(path);
 
-            // Call the OSX or other unix implementation
-            try
-            {
-                SetAccessOrWriteTimeCore(path, time, isAccessTime);
-            }
-            finally
-            {
-                InvalidateCaches();
-            }
-        }
-
-        internal void SetCreationTime(string path, DateTimeOffset time)
-        {
-            // Note: we don't invalidate the cache here since it is not necessary on OSX (and will
-            // reduce performance), it is invalidated on other unix platforms in their method.
-
-            // Call the OSX or other unix implementation
-            try
-            {
-                SetCreationTimeCore(path, time, allowFallbackToLastWriteTime: true);
-            }
-            finally
-            {
-                InvalidateCaches();
-            }
-        }
-
-        private unsafe void SetAccessOrWriteTimeImpl(string path, DateTimeOffset time, bool isAccessTime, bool checkCreationTime)
-        {
             // we use utimes()/utimensat() to set the accessTime and writeTime
             Interop.Sys.TimeSpec* buf = stackalloc Interop.Sys.TimeSpec[2];
 
@@ -368,12 +318,15 @@ namespace System.IO
             // On OSX, when the modification time is less than the creation time, the creation time is broken;
             // so these api calls revert the creation time when it shouldn't be set (since we're setting
             // modification time and access time here). checkCreationTime is only true on OSX-like platforms.
-            if (checkCreationTime && (_fileCache.Flags & Interop.Sys.FileStatusFlags.HasBirthTime) != 0)
+            bool updateCreationTime = checkCreationTime && (_fileCache.Flags & Interop.Sys.FileStatusFlags.HasBirthTime) != 0 &&
+                                        (_fileCache.MTime < _fileCache.BirthTime || (_fileCache.MTime == _fileCache.BirthTime && _fileCache.MTimeNsec < _fileCache.BirthTimeNsec));
+
+            InvalidateCaches();
+
+            if (updateCreationTime)
             {
-                if (_fileCache.MTime < _fileCache.BirthTime || (_fileCache.MTime == _fileCache.BirthTime && _fileCache.MTimeNsec < _fileCache.BirthTimeNsec))
-                {
-                    SetCreationTimeCore(path, UnixTimeToDateTimeOffset(_fileCache.BirthTime, _fileCache.BirthTimeNsec), allowFallbackToLastWriteTime: false);
-                }
+                SetCreationTimeCore(path, UnixTimeToDateTimeOffset(_fileCache.BirthTime, _fileCache.BirthTimeNsec));
+                // TODO: throw if not ENOTSUP and not SUCCESS.
             }
         }
 
```

Reply:
> We should mostly care about the syscalls. Other stuff like inlineability won't be noticeable because the syscalls are far more costly than what we're doing in .NET code.

Sorry, I thought you meant something else with something you said earlier.

I looked through your code and worked out what you were doing. I based my one off of the ideas of yours but wrote it all myself (hence some ordering differences etc). My version manages to avoid the field `s_creationTimeNotSupported` (which may have ended up with concurrency issues, idk).
Let me know what you think of my new version, I think this version is essentially done (other than updating some comments).
[Archive.zip](https://github.com/dotnet/runtime/files/7139992/Archive.zip)

Reply:
Thanks for helping me refactor the code to look and work better!

Reply:
@hamarb123 one thing to consider instead of attaching zips to show possible changes is to just push a branch to your fork for whoever to look at. That makes it easy as it's in GitHub

Reply:
That's a good point @danmoseley, I've made the testing branch over at https://github.com/hamarb123/runtime/tree/testing now and will be pushing commits there until I want it to CI. The commit details will not necessarily be copied as I'll copy the files into the other branch and do it in 1 commit instead, I'll provide a more useful message when I push to hamarb123:main.

Reply:
> My version manages to avoid the field s_creationTimeNotSupported (which may have ended up with concurrency issues, idk).

`s_creationTimeNotSupported` is an optimization for when the underlying system doesn't support `setattrlist`. Then we remember we shouldn't use it again, and avoid the cost of the syscall. I don't know if this is expected to be a common case. It can be left out.

We don't need to worry about concurrency because we end up in the same state eventually.

Reply:
In my code, which syscall is it avoiding; I think I wrote mine in a way to avoid unnecessary syscalls.

Reply:
I'll take a look when you push these changes.

You can mark this discussion as resolved.

Reply:
Okay, I'll push them in a moment. I might not be able to respond to anything much for the next day though, we'll see.